jobs/worker.go

```diff
@@ -1,46 +1,84 @@
 package jobs
 
 import (
+	"context"
+	"errors"
 	"time"
 
 	l "github.com/RedHatInsights/sources-api-go/logger"
+	"github.com/RedHatInsights/sources-api-go/redis"
 )
 
-var ch chan Job
-
-// fire up a worker, initializing the channel to read from inline
-func Init() {
-	// initialize the channel to be used by the job worker. The `RunWorker`
-	// function just takes a channel so we can swap out the backing store for
-	// the jobs for anything we like as long as it can speak through a channel
-	// (maybe redis groups someday.)
-	ch = make(chan Job)
-	l.Log.Infof("Starting Worker for Delayed Jobs")
-	go RunWorker(ch)
+// the queue on redis we'll be sending the jobs to
+const workQueue = "sources_api_jobs"
+
+// Runs the worker just consuming off of a redis list
+func Run() {
+	l.Log.Infof("Starting up Background worker listening to redis queue %q", workQueue)
+
+	for {
+		// This is a BLocking Pop that will effectively wait forever until
+		// something gets queued.
+		//
+		// Once it does succeed - we check to see if the error is just
+		// `redis.Nil`, in which case things are fine and we continue.
+		val, err := redis.Client.BLPop(context.Background(), 0, workQueue).Result()
+		if err != nil {
+			if errors.Is(err, redis.Nil) {
+				l.Log.Debugf("nil value returned  from redis - continuing")
+			} else {
+				l.Log.Warnf("Failed to pop job from queue: %v", err)
+			}
+
+			continue
+		}
+
+		jr := JobRequest{}
+		// the val that is returned is a slice in the form [listname, value]
+		// where val[0] is the name of the list (e.g. workQueue above) and
+		// val[1] is the string value, e.g. the output from
+		// jobRequest.MarshalBinary. So we need to unmarshal it.
+		err = jr.UnmarshalBinary([]byte(val[1]))
+		if err != nil {
+			l.Log.Warnf("Failed to unmarshal job from redis: %v", err)
+			continue
+		}
+
+		// Parse is the _very_ important method that looks at the job name and
+		// figures out what kind of job to unmarshal
+		err = jr.Parse()
+		if err != nil {
+			l.Log.Warnf("Failed to unmarshal job from redis: %v", err)
+			continue
+		}
+
+		RunJobAsync(jr.Job)
+	}
 }
 
-// Package level function to enqueue jobs to the job channel
-func Enqueue(j Job) {
-	ch <- j
+// Run a Job with a delay but in the background so it does _not_ block the
+// calling routine.
+func RunJobAsync(j Job) {
+	l.Log.Infof("Running job asyncronously %v", j.Name())
```

Comment:
What about including the ID here? This way if something goes wrong we know exactly which one failed